Asset CDN / Photon: group the 2 features under a single module card in the admin.

jeherve opened the pull request:
#### Changes proposed in this Pull Request:

- Update the support link to use the new name we've given to the Photon feature.
- Remove mention of "saving space" since Photon does not do that. => Fixes #10300
- Fix a comment: Tiled Galleries depend on Photon, not Carousel.
- **Create a new admin UI card with both the Photon and Asset CDN options, grouped under a single toggle.**

@see p9jf6J-11u-p2 for more information about this. The copy is not final, as per the discussion there.

**Before**

<img width="791" alt="screenshot 2018-10-12 at 15 46 16" src="https://user-images.githubusercontent.com/426388/46873029-f98a9e80-ce35-11e8-900f-2df8aa04e963.png">

**After**

<img width="776" alt="screenshot 2018-10-22 at 11 42 52" src="https://user-images.githubusercontent.com/426388/47286596-a4961780-d5ef-11e8-82fa-8e8ec5a78595.png">


#### Testing instructions:

0. Add `add_filter( 'jetpack_force_disable_site_accelerator', '__return_true' );` to a functionality plugin on your site to avoid getting old JS files when viewing the admin panel with the Asset CDN module on.
1. Go to Jetpack > Settings > Writing.
2. You should see the new section, reflecting the current status of your site.
3. Toggling the main option should allow you to turn everything on at once.
4. You can still toggle the 2 options separately.
5. If your site is in development mode, you cannot toggle the Photon option.

One annoyance you'll face during testing is that when turning on the Asset CDN option, and then refreshing the page, the admin interface will be broken. That is because the Asset CDN feature cannot serve files from a development version at the moment. To turn the option back off, you will need to go to this page:

`{yoursite.com}/wp-admin/admin.php?page=jetpack_modules&s=asset`

#### Proposed changelog entry for your changes:
* Asset CDN: add option to toggle the new feature from the Jetpack Dashboard.

jeherve committed:
Photon: update admin UI description and support link.

Fixes #10300

- Update the support link to use the new name we've given to the Photon feature.
- Remove mention of "saving space" since Photon does not do that.

jeherve committed:
Photon card: Tiled Galleries depend on Photon, not Carousel.

jeherve committed:
Photon: create new admin UI card with both Photon and Asset CDN

@see p9jf6J-11u-p2

jeherve commented on _inc/client/writing/speed-up-site.jsx:
```diff
@@ -30,52 +32,87 @@ const SpeedUpSite = moduleSettingsForm(
+			const currentCdn = this.props.getOptionValue( 'photon-cdn' );
+
+			// Tiled Galleries depends on Photon. Deactivate it when Photon is deactivated.
+			this.props.updateOptions( {
```
I spotted an issue with that logic, when trying to toggle the option off with only one of the modules active. This will need to be updated.

ebinnion commented:
It looks like this should work with module overrides, based on using the `ModuleToggle` component. But, it may be worth testing. You can do that by forcing the module off with the `option_jetpack_active_modules` or `jetpack_active_modules` filters.

After disabling a module with one of those filters, the toggle should be grayed out and you should see a tooltip when you hover.

https://jetpack.com/support/module-overrides/

jeherve commented on _inc/client/writing/speed-up-site.jsx:
```diff
@@ -30,52 +32,87 @@ const SpeedUpSite = moduleSettingsForm(
 							module={ photon }
 							support={ {
-								link: 'https://jetpack.com/support/photon/',
+								link: 'https://jetpack.com/support/image-cdn/',
```
Right now we only have one support link, and I'd be tempted to create one for the 2 options, and then link to the existing docs for more info about each option. I think it may be less confusing for the users that way.

brbrr commented:
It is not possible to disable all the sub-toggles via the main toggle: 
Steps to reproduce:
- have Image CDN enabled
- in Jetpack Beta select this PR
- Go to Settings > Writing
- Click on `Enable Site Accelerator` toggle

![screencast 2018-10-12 18-08-43](https://user-images.githubusercontent.com/5654161/46877795-542df580-ce4a-11e8-81b1-dde398c28386.gif)

jeherve committed:
Photon: fix toggle handling when one option is on and the other off

@see https://github.com/Automattic/jetpack/pull/10311#issuecomment-429359868

jeherve committed:
Photon toggle: handle potential module overrides.

While the individual module toggles' appearance is automatically managed,
the toggle that allows one to toggle both modules at once is not.
This commit adds the necessary checks so one can't use the main toggle
to change the status of a module when it is forbidden.

jeherve commented:
@ebinnion Thanks for the warning. While the individual toggles worked well with the module overrides, I needed to handle the behaviour of the new overall toggle that can allow one to enable the 2 modules at once. I took care of that in b6f7f78560e0c26c91ede34f977e1784c1596826

ebinnion commented:
I think we'll want to change the logic slightly:

<img width="749" alt="screen shot 2018-10-15 at 8 59 11 am" src="https://user-images.githubusercontent.com/1126811/46955577-da1f8b80-d058-11e8-9db6-5c505ed8c8b2.png">

In this case I've forced on Photon and forced off the Photon CDN, but the toggle to control both doesn't appear disabled.

I think changing to something like this might handle it:

```
// Is at least one of the 2 modules available (not hidden via a module override)?
const photonStatus = this.props.getModuleOverride( 'photon' );
const photonCdnStatus = this.props.getModuleOverride( 'photon-cdn' );
const canDisplayCdnSettings = ( 'inactive' !== photonStatus || 'active' !== photonStatus ) && ( 'inactive' !== photonCdnStatus || 'active' !== photonCdnStatus )
```

jeherve commented:
> In this case I've forced on Photon and forced off the Photon CDN, but the toggle to control both doesn't appear disabled.

Should it be disabled? You are still able to turn on the feature we promote as "Site Accelerator", you're just not getting all of it since a site admin disabled part of it. If we mark that main toggle as disabled, wouldn't it cause confusion for some people, who would assume they can't use any of the "sub" features because the master toggle is disabled?

ebinnion commented:
> Should it be disabled?

Good question. I'm not 100% sure. I guess the question is, if the user were to toggle the "Site Accelerator" toggle, would they expect that both the photon and photon CDN toggles be toggled as well? My initial assumption was yes, but I'm not sure now.

rickybanister commented:
This looks great for the initial release. We are looking to make everything slightly simpler going forward, but a master toggle the flips on image and file service is our best bet. Let's ship it.

rickybanister commented:
I discussed the issue of module overrides with @jeherve a bit and I believe it's fine to ship in the current state (@ebinnion ping me if you want to chat through it). I think it will work alright if one sub toggle is disabled and the other is enabled—the parent toggle will simply be an auxiliary switch for the enabled sub toggle. If a customer is confused as to why they can't turn on say, static files, they'll hover and learn that it's been disabled. That works fine in my eyes and requires less to maintain in the UI than the other solutions.

rickybanister commented:
The only copy suggestion I have is to make `site accelerator` lower case since it's not a proper noun or a trademark. We can consider the feature to be an accelerator of your site, but not a *Site Accelerator*. 

@Automattic/editorial for a gut check please!

jeherve committed:
Asset CDN: Site Accelerator should be lowercase.

@see https://github.com/Automattic/jetpack/pull/10311#issuecomment-430644874

jeherve commented:
> The only copy suggestion I have is to make site accelerator lower case since it's not a proper noun or a trademark.

Done in 231d0df03dcd7e5a3202d3e10d7209d121f2231e

For anyone looking to view the text on a test site quickly, you can use this link:
https://jurassic.ninja/create?jetpack-beta&branch=add/asset-cdn-interface&shortlived&wp-debug-log&gutenberg

It will launch a test site where you can connect to WordPress.com, and then go to Jetpack > Settings > Writing in wp-admin to view the updated "Performance & speed" card.

michelleweber commented:
Gut, checked! If it's not a proper noun/a formal feature name, lowercase.

jeherve committed:
Speed Up Site: simplify check if one of the modules is active

jeherve committed:
Admin Page: make sure the main toggle always appears in Search results.

jeherve committed:
Site Accelerator: track clicks and events when using main toggle

@see https://github.com/Automattic/jetpack/pull/10311#issuecomment-430795435

jeherve committed:
Asset CDN: add new filter to short-circuit the module.

It allows one to easily test any changes in Jetpack, even when the Asset CDN is on,
without being served assets from the CDN.

jeherve committed:
Site accelerator: add toggle status to track click event.

jeherve committed:
Site Accelerator: handle toggling status for the main toggle.

jeherve committed:
Site Accelerator: rename variables to avoid confusion

Photon, Photon CDN, Asset CDN, CDN, it all gets a bit confusing. Hopefully this will help:
- Photon = Photon, the module
- Asset CDN = the module previously known as Photon CDN.
- Site Accelerator: the main toggle. Not a module per se, but a way to turn on the 2 modules at once.

jeherve committed:
Site Accelerator: add new support doc link.

jeherve committed:
Site Accelerator: update wording based on feedback from Editorial

@see https://github.com/Automattic/wp-calypso/pull/27944#issuecomment-431178239

keoshi reviewed (approved):
LGTM!

keoshi reviewed (approved):
LGTM!

keoshi reviewed (dismissed):
LGTM!

keoshi commented:
Apologies for the repeated reviews above, Github was acting up this morning.

jeherve committed:
Asset CDN: add module info to filter docblocks

joanrho reviewed (dismissed):
Let's change

>Lazy-loading images improve your site's speed and create a smoother viewing experience. Images will load as visitors scroll down the screen, instead of all at once.

to 

>Lazy-loading images improves your site's speed and creates a smoother viewing experience. Images will load as visitors scroll down the screen, instead of all at once.

because the subject of the sentence is "lazy-loading", not images.

Other than that, LGTM!

benhuberman commented:
@joanrho I think it's very easy to read "lazy-loading" as modifying "images" rather than as present participle -- that's how I'd instinctively read it -- so I'd actually keep the current agreement in place. If you sense it's too ambivalent as-is, you could reword a bit more, e.g.:
> When images lazy-load, your site's speed improves and the viewing experience becomes smoother.

joanrho commented:
@benhuberman How about something like this?

>Lazy-loading images will improve your site’s speed and create a smoother viewing experience. Images will load as visitors scroll down the screen, instead of all at once.

rickybanister commented:
I agree with your final edit there @joanrho, otherwise I would have sided with Ben on 'images' being the  subject. Nice fix though.

benhuberman commented:
> How about something like this?

I really like your edit, @joanrho!

joanrho committed:
Lazy Images: update module description in admin UI.

@see https://github.com/Automattic/jetpack/pull/10311#issuecomment-432438781

jeherve commented:
👍 	Done in 74bf3159fcd91c74e1f7444fc516b2a2127f5c4a

kraftbj reviewed (approved):
Works as advertised! One question for my own learning, but works well. 🚢
